iroha/src/p2p.rs

```diff
@@ -0,0 +1,259 @@
+            Ok(Response::new())
+        };
+
+        async fn handle_connection(
+            state: State<NodeState>,
+            stream: Box<dyn AsyncStream>,
+        ) -> Result<(), String> {
+            Network::handle_message_async(state, stream, handle_request).await
+        };
+
+        let listen_address = self.state.lock().await.listen_address.clone();
+        Network::listen_async(
+            self.state.clone(),
+            listen_address.as_ref(),
+            handle_connection,
+        )
+        .await
+    }
+
+    pub async fn start_and_connect(&self, peer_address: &str) -> Result<(), String> {
+        let peer = Peer {
+            listen_address: peer_address.to_string(),
+        };
+        self.state.lock().await.peers.insert(peer.clone());
+        let message = Message::NewPeer(Peer {
+            listen_address: self.state.lock().await.listen_address.clone(),
+        });
+        Self::send(&message, peer).await?;
+        self.start().await?;
+        Ok(())
+    }
+
+    async fn send(message: &Message, peer: Peer) -> Result<(), String> {
+        let _response = Network::send_request_to(
+            peer.listen_address.as_ref(),
+            Request::new("/".to_string(), message.into()),
+        )?;
+        Ok(())
+    }
+
+    async fn broadcast(state: State<NodeState>, message: &Message) -> Result<(), String> {
+        for peer in state.lock().await.peers.clone() {
+            Self::send(message, peer).await?;
+        }
+        Ok(())
+    }
+
+    async fn ping(state: State<NodeState>, peer: Peer) -> Result<(), String> {
+        let ping = Ping {
+            payload: gen_random_bytes(PING_SIZE),
+            to_peer: peer.clone(),
+            from_peer: Peer {
+                listen_address: state.lock().await.listen_address.clone(),
+            },
+        };
+        state
+            .lock()
+            .await
+            .sent_pings
+            .insert(ping.clone(), current_time());
+        Self::send(&Message::Ping(ping), peer.clone()).await
+    }
+
+    async fn ping_all(state: State<NodeState>) -> Result<(), String> {
+        for peer in state.lock().await.peers.clone() {
+            Self::ping(state.clone(), peer).await?;
+        }
+        Ok(())
+    }
+
+    async fn handle_message(state: State<NodeState>, bytes: &[u8]) -> Result<(), String> {
+        let message: Message = bytes.to_vec().into();
+        match message {
+            Message::Ping(ping) => {
+                Self::send(&Message::Pong(ping.clone()), ping.from_peer).await?;
+            }
+            Message::Pong(ping) => {
+                let sent_pings = &mut state.lock().await.sent_pings;
+                if sent_pings.contains_key(&ping) {
+                    let sent_time = sent_pings
+                        .get(&ping)
+                        .expect("Failed to get sent ping entry.");
+                    let _rtt = current_time() - sent_time.to_owned();
```

Comment:
`Duration::since` can be used and private function current_time can be removed.
Utility function are bad for projects and let's try to use them only in tests.